lib/datadog/core/telemetry/collector.rb

```diff
@@ -0,0 +1,196 @@
+          tracer_version if Datadog.configuration.respond_to?(:profiling) && Datadog.configuration.profiling.enabled
+        end
+
+        def appsec
+          version = appsec_version
+          Telemetry::V1::AppSec.new(version: version) if version
+        end
+
+        def appsec_version
+          tracer_version if Datadog.configuration.respond_to?(:appsec) && Datadog.configuration.appsec.enabled
+        end
+
+        def configuration_variables
+          configurations = []
+          [
+            Telemetry::V1::Configuration.new(name: 'diagnostics.debug', value: Datadog.configuration.diagnostics.debug),
+            Telemetry::V1::Configuration.new(name: 'diagnostics.startup_logs.enabled',
+                                             value: Datadog.configuration.diagnostics.startup_logs.enabled),
+            Telemetry::V1::Configuration.new(name: 'profiling.advanced.endpoint.collection.enabled',
+                                             value: Datadog.configuration.profiling.advanced.endpoint.collection.enabled),
+            Telemetry::V1::Configuration.new(name: 'profiling.advanced.code_provenance_enabled',
+                                             value: Datadog.configuration.profiling.advanced.code_provenance_enabled),
+            Telemetry::V1::Configuration.new(name: 'profiling.advanced.legacy_transport_enabled',
+                                             value: Datadog.configuration.profiling.advanced.legacy_transport_enabled),
+            Telemetry::V1::Configuration.new(name: 'runtime_metrics.enabled',
+                                             value: Datadog.configuration.runtime_metrics.enabled),
+            Telemetry::V1::Configuration.new(name: 'tracing.enabled', value: Datadog.configuration.tracing.enabled),
+            Telemetry::V1::Configuration.new(name: 'tracing.analytics.enabled',
+                                             value: Datadog.configuration.tracing.analytics.enabled)
+          ].each do |configuration|
+            configurations << configuration unless configuration.value.nil?
+          end
+          configurations
+        end
+
+        def instrumented_integrations
+          Datadog.configuration.tracing.instrumented_integrations
+        end
+
+        def instrumented?(integration)
+          instrumented_integrations.include?(integration.name)
+        end
+
+        def patched?(integration)
+          instrumented_integrations[integration.name].patch == true
+        end
+
+        def integration_auto_instrument?(integration)
+          integration.klass.auto_instrument?
+        end
+
+        def integration_compatible?(integration)
+          integration.klass.class.compatible?
+        end
+
+        def integration_version(integration)
+          integration.klass.class.version ? integration.klass.class.version.to_s : nil
+        end
+
+        def patch_result(integration)
+          instrumented_integrations[integration.name].patch
```

Comment:
Calling `#patch` has the side effect of actually patching the host application, even if the user doesn't want to.
I don't think that's the intent here.

I noticed the result of patch is used to create the `def integration_error` below.

So we are looking at the result of patch in case of failure.
Because we can't forcefully patch just to figure that out, have have to use the last result of `#patch` calls.

We don't store the outcome of `#patch` today.

`#patch` is currently wrapped around a `patch_only_once`: https://github.com/DataDog/dd-trace-rb/blob/265fa48a736527241265c07a768f9883679eed30/lib/datadog/tracing/contrib/patcher.rb#L33-L42
This means it only runs once, even if called many times.

I think the best course here is to store the result of `#patch` in the https://github.com/DataDog/dd-trace-rb/blob/265fa48a736527241265c07a768f9883679eed30/lib/datadog/tracing/contrib/patcher.rb class (which will store it in each instrumentation class), so we can inspect it later (later being right here in the collector).

We'd then call `integration.klass.class.latest_patch_error` and that would be it.